src/cpu/pred/btb.hh

```diff
@@ -29,103 +41,87 @@
 #ifndef __CPU_PRED_BTB_HH__
 #define __CPU_PRED_BTB_HH__
 
+
 #include "arch/generic/pcstate.hh"
-#include "base/logging.hh"
-#include "base/types.hh"
+#include "base/statistics.hh"
+#include "cpu/pred/branch_type.hh"
+#include "cpu/static_inst.hh"
+#include "params/BranchTargetBuffer.hh"
+#include "sim/clocked_object.hh"
 
 namespace gem5
 {
 
 namespace branch_prediction
 {
 
-class DefaultBTB
+class BranchTargetBuffer : public ClockedObject
 {
-  private:
-    struct BTBEntry
-    {
-        /** The entry's tag. */
-        Addr tag = 0;
-
-        /** The entry's target. */
-        std::unique_ptr<PCStateBase> target;
+  public:
+    typedef BranchTargetBufferParams Params;
+    typedef enums::BranchType BranchType;
 
-        /** The entry's thread id. */
-        ThreadID tid;
+    BranchTargetBuffer(const Params &params);
 
-        /** Whether or not the entry is valid. */
-        bool valid = false;
-    };
+    virtual void memInvalidate() override = 0;
 
-  public:
-    /** Creates a BTB with the given number of entries, number of bits per
-     *  tag, and instruction offset amount.
-     *  @param numEntries Number of entries for the BTB.
-     *  @param tagBits Number of bits for each tag in the BTB.
-     *  @param instShiftAmt Offset amount for instructions to ignore alignment.
+    /** Looks up an address in the BTB. Upon hit returns the target of the
+     * branch. Returns nullptr.
+     *  @param inst_PC The address of the branch to look up.
+     *  @param type Optional type of the branch to look up.
+     *  @return The target of the branch or nullptr if the branch is not
+     *          in the BTB.
      */
-    DefaultBTB(unsigned numEntries, unsigned tagBits,
-               unsigned instShiftAmt, unsigned numThreads);
+    virtual const PCStateBase *lookup(ThreadID tid, Addr instPC,
+                            BranchType type = BranchType::NoBranch) = 0;
 
-    void reset();
-
-    /** Looks up an address in the BTB. Must call valid() first on the address.
+    /** Looks up an address in the BTB and return the instruction
```

Comment:
It might be worth renaming this to `getInst` or `lookupInstNoStats` to indicate that this doesn't actually update any stats. You should also document this behaviour.